Storage chain fixes + guide (#9504)
paritytech/substrate: frame/transaction-storage/README.md

# Transaction Storage Pallet

Indexes transactions and manages storage proofs.

> Allows storing arbitrary data on the shain. Data is automaticcaly removed after `StoragePeriod` blocks, unless the storage is renewed.

Reviewer: ```suggestion
Allows storing arbitrary data on the chain. Data is automatically removed after `StoragePeriod` blocks, unless the storage is renewed.
```

Validators must submit proof of storing a random chunk of data for block `N - StoragePeriod` when producing block `N`.

> # Runing a chain

Reviewer: ```suggestion
# Running a chain
```

> The following describes how to setup a new storage chain.

Reviewer: ```suggestion
The following describes how to set up a new storage chain.
```

Start with generating a chain spec.

> ```

Reviewer: ```suggestion

```bash
```

cargo run --release -- build-spec --chain=local > sc_init.json
```

> Edit the json chain spec file to customise the chain. The storage chain genesis params are configured in the `transactionStorage` section.

Reviewer: ```suggestion

Edit the json chain spec file to customise the chain. The storage chain genesis params are configured in the `transactionStorage` section.
```

> Note that `storagePeriod` is specified in blocks and changing it also requires code changes at the moment.

Reviewer 2: Could be good to have a sample of chain spec with the specified sections.

Author: Generated default is the sample. All the values are there.

Build a raw spec from the init spec.

> ```

Reviewer: ```suggestion

```bash
```

cargo run --release build-spec --chain=sc_init.json --raw > sc.json
```

> Run a few validator nodes.

Reviewer: ```suggestion

Run a few validator nodes.
```

> ```

Reviewer: ```suggestion
```bash
```

cargo run --release -- --chain=sc.json -d /tmp/alice --storage-chain --keep-blocks=100800 --ipfs-server --validator --alice
cargo run --release -- --chain=sc.json -d /tmp/bob --storage-chain --keep-blocks=100800 --ipfs-server --validator --bob
```

`--storage-chain` enables transaction indexing.
`--keep-blocks=100800` enables block pruning. The value here should be greater or equal than the storage period.
`--ipfs-server` enables serving stored content over IPFS.

> Once the network is started, any other joining nodes need to sync with `--sync=fast`. Regular sync will fail because block pruning removes old blocks. The chain does noot keep full block history.

Reviewer: ```suggestion
Once the network is started, any other joining nodes need to sync with `--sync=fast`. Regular sync will fail because block pruning removes old blocks. The chain does not keep full block history.
```

> ```

Reviewer: ```suggestion
```bash
```

cargo run --release -- --chain=sc.json -d /tmp/charlie --storage-chain --keep-blocks=100800 --ipfs-server --validator --charlie --sync=fast
```

> # Making transactions

Reviewer: ```suggestion

# Making transactions
```

To store data use the `transactionStorage.store` extrinsic. And IPFS CID can be generated from the Blake2-256 hash of the data.

```js
	const file = fs.readFileSync('cute_kitten.jpeg');
	const hash = util_crypto.blake2AsU8a(file)
	const encoded_hash = multihash.encode(hash, 'blake2b-256');
	const cid = new CID(1, 'blake2b-256', encoded_hash)
	console.log(cid.toString());
	const txHash = await api.tx.transactionStorage.store('0x' + file.toString('hex')).signAndSend(alice);

> ```

Reviewer: ```suggestion
```

```

Author: This suggestion does not look righ

Data can be queried over IPFS

> ```

Reviewer: ```suggestion
```bash
```

Build a raw spec from the init spec.

```bash
cargo run --release build-spec --chain=sc_init.json --raw > sc.json
```

Run a few validator nodes.

```bash
cargo run --release -- --chain=sc.json -d /tmp/alice --storage-chain --keep-blocks=100800 --ipfs-server --validator --alice
cargo run --release -- --chain=sc.json -d /tmp/bob --storage-chain --keep-blocks=100800 --ipfs-server --validator --bob
```

`--storage-chain` enables transaction indexing.
`--keep-blocks=100800` enables block pruning. The value here should be greater or equal than the storage period.
`--ipfs-server` enables serving stored content over IPFS.

Once the network is started, any other joining nodes need to sync with `--sync=fast`. Regular sync will fail because block pruning removes old blocks. The chain does not keep full block history.

```bash
cargo run --release -- --chain=sc.json -d /tmp/charlie --storage-chain --keep-blocks=100800 --ipfs-server --validator --charlie --sync=fast
```

# Making transactions

To store data use the `transactionStorage.store` extrinsic. And IPFS CID can be generated from the Blake2-256 hash of the data.

```js

> 	const file = fs.readFileSync('cute_kitten.jpeg');

Reviewer 3: @arkpar Could we have a more complete example, e.g., include the snippets of js imports? So that we can use the js example directly.

Author: Expanded it a bit
